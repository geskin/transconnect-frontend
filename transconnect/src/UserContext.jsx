import { createContext, useState, useEffect } from "react";

const UserContext = createContext();

export const UserProvider = ({ children }) => {
    const [currUser, setCurrUser] = useState(() => {
        const storedUser = localStorage.getItem("currUser");
        return storedUser ? JSON.parse(storedUser) : null;
    });

    // Save to localStorage whenever currUser changes
    useEffect(() => {
        if (currUser) {
            localStorage.setItem("currUser", JSON.stringify(currUser));
        } else {
            localStorage.removeItem("currUser");
        }
    }, [currUser]);

    return (
        <UserContext.Provider value={{ currUser, setCurrUser }}>
            {children}
        </UserContext.Provider>
    );
};

export default UserContext;

