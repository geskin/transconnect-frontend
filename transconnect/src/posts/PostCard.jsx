const PostCard = ({
    title,
    id,
    createdAt,
    editedAt,
    content,
    user,
    comments }) => {

    return (
        <Link className="card" to={`/posts/${id}`}>
            <div className="card-body">
                <h3>{user.username}</h3>
                <h6 className="card-title text-start">
                    {title}
                </h6>
                <p className="text-start"><small>{content}</small></p>
                <i>{createdAt}</i>
            </div>
        </Link>
    );
}

export default PostCard;