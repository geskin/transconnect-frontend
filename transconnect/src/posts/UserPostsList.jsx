import React, { useContext, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import UserContext from "../UserContext";
import TransconnectApi from "../api";
import PostCard from "./PostCard";

/** UserPostsList: list of posts (in card format) made by a specific user
 * 
 * auth required: logged in user
 */

const UserPostsList = () => {
    const { username } = useParams();
    const { currUser } = useContext(UserContext);
    const [user, setUser] = useState({});
    const [posts, setPosts] = useState([]);

    useEffect(() => {
        if (!currUser) return;

        const fetchUser = async () => {
            try {
                let user = await TransconnectApi.getUser(username);
                setUser(user)
            } catch (err) {
                console.error(`Error fetching user: ${username}`, err);
            }
        }

        const fetchUserPosts = async () => {
            try {
                let userPosts = await TransconnectApi.getUserPosts(user.id);
                setPosts(userPosts);
            } catch (err) {
                console.error(`Error fetching ${username}'s posts`, err);
            }
        }

        fetchUser();
        fetchUserPosts();
    });

    const handleDeletePost = async (postId) => {
        try {
            await TransconnectApi.deletePost(postId);
            setPosts(posts.filter(post => post.id !== postId));
        } catch (err) {
            console.error("Error deleting post", err);
        }
    };

    return (
        <div>
            <div>
                {posts.length > 0 ? (
                    posts.map((p) => (
                        <PostCard
                            key={p.id}
                            id={p.id}
                            title={p.title}
                            createdAt={p.createdAt}
                            editedAt={p.editedAt}
                            content={p.content}
                            user={p.user}
                            comments={p.comments}
                            tags={p.tags}
                            onDelete={handleDeletePost}
                        />
                    ))
                ) : (
                    <p className="lead white-letters">Sorry, no results were found!</p>
                )}
            </div>
        </div>
    );

}

export default UserPostsList;