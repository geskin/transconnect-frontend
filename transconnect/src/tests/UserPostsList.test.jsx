import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import UserPostsList from "../posts/UserPostsList";
import UserContext from "../UserContext";
import TransconnectApi from "../api";
import PostCard from "../posts/PostCard";
import { MemoryRouter, Route, Routes } from "react-router-dom";

vi.mock("../api");
vi.mock("../posts/PostCard", () => ({
    __esModule: true,
    default: vi.fn(() => <div data-testid="post-card">PostCard</div>),
}));

describe("UserPostsListc component", () => {
    const mockCurrUser = { username: "testuser", id: 1, role: "USER" };
    const mockUser = { username: "otheruser", id: 2 };
    const mockPosts = [
        {
            id: 101,
            title: "First Post",
            createdAt: "2024-03-14T12:00:00Z",
            editedAt: "2024-03-15T12:00:00Z",
            content: "This is a test post",
            user: { username: "otheruser" },
            comments: [],
            tags: [],
        },
    ];

    beforeEach(() => {
        TransconnectApi.getUser.mockResolvedValue(mockUser);
        TransconnectApi.getUserPosts.mockResolvedValue(mockPosts);
    });

    it("renders without crashing", () => {
        render(
            <MemoryRouter initialEntries={["/users/otheruser/posts"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/users/:username/posts" element={<UserPostsList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <MemoryRouter initialEntries={["/users/otheruser/posts"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/users/:username/posts" element={<UserPostsList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("fetches and displays user posts", async () => {
        render(
            <MemoryRouter initialEntries={["/users/otheruser/posts"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/users/:username/posts" element={<UserPostsList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByTestId("post-card")).toBeInTheDocument();
        });
    });

    it("displays message when no posts are found", async () => {
        TransconnectApi.getUserPosts.mockResolvedValue([]);

        render(
            <MemoryRouter initialEntries={["/users/otheruser/posts"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/users/:username/posts" element={<UserPostsList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByText("Sorry, no results were found!")).to.exist;
        });
    });

    it("handles post deletion", async () => {
        TransconnectApi.deletePost.mockResolvedValue({});
        render(
            <MemoryRouter initialEntries={["/users/otheruser/posts"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/users/:username/posts" element={<UserPostsList />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByTestId("post-card")).to.exist;
        });

        // Simulate deleting a post
        fireEvent.click(screen.getByTestId("post-card")); // Simulate clicking delete on a post card

        await waitFor(() => {
            expect(screen.queryByTestId("post-card")).to.not.exist;
        });
    });
});
