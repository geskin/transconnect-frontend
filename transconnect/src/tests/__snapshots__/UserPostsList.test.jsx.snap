// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserPostsList > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <p
        class="lead white-letters"
      >
        Sorry, no results were found!
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`UserPostsListc component > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <p
        class="lead white-letters"
      >
        Sorry, no results were found!
      </p>
    </div>
  </div>
</DocumentFragment>
`;
